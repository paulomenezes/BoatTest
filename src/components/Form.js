import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { reduxForm, Field } from 'redux-form';

import styles from './Form.styles';
import TextEdit from './TextEdit';
import TextEditIcon from './TextEditIcon';

export const required = value => (value ? undefined : true);
export const minLength = min => value =>
  value && value.length < min ? true : undefined;
export const length6 = value =>
  value && value.length !== 6 ? true : undefined;
export const alphaNumeric = value =>
  value && /[^a-zA-Z0-9 ]/i.test(value) ? true : undefined;
const minLength2 = minLength(2);
const minLength3 = minLength(3);

const Form = props => (
  <View style={styles.container}>
    <View style={styles.fieldRow}>
      <Field
        style={styles.inputLeft}
        name="bookingReference"
        component={TextEdit}
        placeholder="Booking Reference"
        validate={[required, alphaNumeric, length6]}
      />
      <Field
        style={styles.inputRight}
        name="lastName"
        component={TextEdit}
        placeholder="Last name"
        validate={[required, minLength2]}
      />
    </View>
    <View style={styles.input}>
      <Field
        name="departing"
        component={TextEditIcon}
        placeholder="Departing"
        validate={[required, minLength3]}
      />
    </View>
    <TouchableOpacity style={styles.button}>
      <View style={styles.buttonShadow}>
        <Text style={styles.buttonText}>Find booking</Text>
      </View>
    </TouchableOpacity>
  </View>
);

export default reduxForm({ form: 'form' })(Form);
