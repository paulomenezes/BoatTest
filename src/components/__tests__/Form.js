import 'react-native';
import React from 'react';
import { reduxForm } from 'redux-form';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import Form, { required, minLength, length6, alphaNumeric } from '../form.js';

import renderer from 'react-test-renderer';

jest.mock('../TextEdit', () => 'TextEdit');
jest.mock('../TextEditIcon', () => 'TextEditIcon');

const store = createStore(() => ({}));
const Decorated = reduxForm({ form: 'form' })(Form);

it('renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Decorated />
    </Provider>
  );

  expect(tree).toMatchSnapshot();
});

describe('validations', () => {
  it('Required validation', () => {
    expect(required('KY5VHA')).toBeFalsy();
    expect(required('Smith')).toBeFalsy();
    expect(required('SIN')).toBeFalsy();
    expect(required(null)).toBeTruthy();
    expect(required(undefined)).toBeTruthy();
    expect(required('')).toBeTruthy();
  });

  it('Minimum length validation', () => {
    expect(minLength(3)('A')).toBeTruthy();
    expect(minLength(3)('AB')).toBeTruthy();
    expect(minLength(3)('ABC')).toBeFalsy();

    expect(minLength(2)('A')).toBeTruthy();
    expect(minLength(2)('AB')).toBeFalsy();
    expect(minLength(2)('ABC')).toBeFalsy();
  });

  it('Character with length 6', () => {
    expect(length6('AAAAA')).toBeTruthy();
    expect(length6('AAAAAA')).toBeFalsy();
  });

  it('Character must be alphanumeric', () => {
    expect(alphaNumeric('ABC')).toBeFalsy();
    expect(alphaNumeric('ABC#')).toBeTruthy();
  });
});
