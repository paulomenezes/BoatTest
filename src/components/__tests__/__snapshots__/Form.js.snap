// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 20,
      }
    }
  >
    <TextEdit
      name="bookingReference"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Booking Reference"
      style={
        Object {
          "marginRight": 5,
        }
      }
      value=""
    />
    <TextEdit
      name="lastName"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Last name"
      style={
        Object {
          "marginLeft": 5,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "height": 40,
        "marginBottom": 20,
      }
    }
  >
    <TextEditIcon
      name="departing"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Departing"
      value=""
    />
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 4,
        "opacity": 1,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 3,
          "width": 3,
        },
        "shadowOpacity": 1,
        "shadowRadius": 0,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Object {
          "backgroundColor": "#000",
          "borderRadius": 4,
          "shadowColor": "#FFF",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
            "fontFamily": "OpenSans-Bold",
            "margin": 10,
            "textAlign": "center",
          }
        }
      >
        Find booking
      </Text>
    </View>
  </View>
</View>
`;
